src/libostree/ostree-repo-prune.c

```diff
@@ -515,10 +567,23 @@ ostree_repo_prune_from_reachable (OstreeRepo        *self,
     return FALSE;
 
   g_autoptr(GHashTable) objects = NULL;
-
-  if (!ostree_repo_list_objects (self, OSTREE_REPO_LIST_OBJECTS_ALL | OSTREE_REPO_LIST_OBJECTS_NO_PARENTS,
+  OstreeRepoPruneFlags flags = options->flags;
+  gboolean list_refs_only = flags & OSTREE_REPO_PRUNE_FLAGS_LIST_REFS_ONLY;
+  gboolean commit_only = flags & OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY;
+  if (list_refs_only) 
+    {
+       OstreeRepoCommitTraverseFlags traverse_flags = 0;
+      if (commit_only)
+        traverse_flags |= OSTREE_REPO_COMMIT_TRAVERSE_FLAG_COMMIT_ONLY;
+      if (!get_objects_from_refs_only (self, traverse_flags, &objects, cancellable, error))
+        return FALSE;      
+    } 
+  else
+    {
+      if (!ostree_repo_list_objects (self, OSTREE_REPO_LIST_OBJECTS_ALL | OSTREE_REPO_LIST_OBJECTS_NO_PARENTS,
```

Comment:
It seems to me that the much safer thing to do is to use `ostree_repo_list_commit_objects_starting_with` with `OSTREE_REPO_PRUNE_FLAGS_COMMIT_ONLY` and `ostree_repo_list_objects` otherwise. You're still scanning all the objects, but the set of objects is much smaller, which will significantly reduce memory pressure.

I feel like the idea that you can consider the full set of objects by only looking at refs is bound to cause corruption sooner or later.

Reply:
I was curious about `ostree_repo_list_objects` vs `ostree_repo_list_commit_objects_starting_with`. At least on our server with our production OS repo (as well as my laptop with `/var/lib/flatpak/repo`), it makes a big difference.

[Here](https://gist.github.com/dbnicholson/4c1898deae258b07d207aa7254ac51e0) is a test program that simply does the scan and prints the number objects it found. The results were surprising:

```
$ time ~/count-objects
Number of objects: 1835291

real	0m9.183s
user	0m7.850s
sys	0m1.332s
$ time ~/count-objects --commits
Number of commits: 660

real	0m1.095s
user	0m0.248s
sys	0m0.847s
```

That's a 9x speedup by throwing out all but the commit objects from the scan. I ran that a few times to make sure it was repeatable. So, I really think you should just try that before introducing new flags.

Reply:
> @dbnicholson Thanks, that's instructive. We're going to try that same test on the Fedora repos to see how it compares and report back.

Here is the script:

```
$ cat ~/script.sh 
cd /mnt/fedora_koji_prod/koji/ostree/repo/
time ~/count-objects
time ~/count-objects --commits

cd /mnt/fedora_koji_prod/koji/compose/ostree/repo/
time ~/count-objects
time ~/count-objects --commits
```

and here are the results:

```
[dustymabe@compose-x86-01 ~][STG]$ bash -x ~/script.sh 
+ cd /mnt/fedora_koji_prod/koji/ostree/repo/
+ /home/fedora/dustymabe/count-objects
Number of objects: 22243384

real    2m32.035s
user    2m15.591s
sys     0m15.070s
+ /home/fedora/dustymabe/count-objects --commits
Number of commits: 25760

real    0m7.071s
user    0m1.392s
sys     0m5.489s
+ cd /mnt/fedora_koji_prod/koji/compose/ostree/repo/
+ /home/fedora/dustymabe/count-objects
Number of objects: 6579571

real    30m12.614s
user    0m37.902s
sys     0m8.075s
+ /home/fedora/dustymabe/count-objects --commits
Number of commits: 9006

real    0m1.752s
user    0m0.391s
sys     0m1.215s
```

> 
> One thing worth pointing out is that these repos have never been pruned so far (this is what we're working on fixing!). So the first prune will be extremely expensive. But once we're over that hump, subsequent prunes should be much cheaper.

One correction here. We did prune the "compose" repo a few months back and dropped out about 2TB of objects/commits. We went ahead and did this repo back then because we have a simple time based prune for the whole repo and so we could prune in a single run and we needed to save disk space because we were running out.

Reply:
I think the `--commits` results there are probably bring overstated because the directory cache is warm, but I'm pretty sure you're going to see a noticeable difference regardless.

Reply:
> One thing worth pointing out is that these repos have never been pruned so far (this is what we're working on fixing!). So the first prune will be extremely expensive. But once we're over that hump, subsequent prunes should be much cheaper.

I was in this situation a few years ago at Endless pre-repo-locking but now we have a daily prune that looks like it took 5 minutes in our worst case repo yesterday. I feel you're pain, though. The initial prune took a really long time after having years of commits piled up in it.